drivers/hv/mshv_vtl_main.c

```diff
@@ -1870,6 +1870,76 @@ static long mshv_vtl_ioctl_guest_vsm_vmsa_pfn(void __user *user_arg)
 }
 #endif
 
+// TODO right ifdefs
+static void ack_kick(void *cancel_cpu_run)
+{
+	// BUGBUG needs to cancel current run right like mshv_vtl_cancel to bring into usermode?
+	// BUGBUG complains "using smp_processor_id() in preemptible code"
+	int cpu;
+	bool cancel = (bool)cancel_cpu_run;
+	cpu = get_cpu();
+	pr_err("CPU %d acked kick\n", cpu);
+
+	if (cancel)
+		WRITE_ONCE(mshv_vtl_this_run()->cancel, 1);
+
+	put_cpu();
+}
+
+static int get_user_cpu_mask(void __user *user_mask_ptr, unsigned len,
+			     struct cpumask *new_mask)
+{
+	if (len < cpumask_size())
+		cpumask_clear(new_mask);
+	else if (len > cpumask_size())
+		len = cpumask_size();
+
+	pr_err("cpu mask len: %d\n", len);
+
+	return copy_from_user(new_mask, user_mask_ptr, len) ? -EFAULT : 0;
+}
+
+static inline long mshv_vtl_ioctl_kick_cpu(void __user *user_arg)
+{
+	struct mshv_kick_cpus args = {};
+	struct cpumask cpus = {};
+	long ret;
+	bool wait_for_cpus = false;
+	bool cancel_cpu_run = false;
+
+	pr_err("kicking cpus requested\n");
+
+	ret = copy_from_user(&args, user_arg, sizeof(args)) ? -EFAULT : 0;
+	if (ret)
+		return ret;
+
+	// read the cpu mask from the user with the specified len
+	ret = get_user_cpu_mask((void __user *)args.cpu_mask_ptr, args.len, &cpus);
+	if (ret)
+		return ret;
+
+
+	if (cpumask_empty(&cpus))
+		return 0;
+
+	if (args.flags & MSHV_KICK_CPUS_FLAG_WAIT_FOR_CPUS)
+		wait_for_cpus = true;
+
+	if (args.flags & MSHV_KICK_CPUS_FLAG_CANCEL_CPU_RUN)
+		cancel_cpu_run = true;
+
+	pr_err("kicking following cpus %*pbl\n", cpumask_pr_args(&cpus));
+
+	preempt_disable();
+	int cpu;
+	cpu = smp_processor_id();
+	pr_err("CPU %d starting kick, wait for cpus: %d, cancel cpu run: %d\n", cpu, wait_for_cpus, cancel_cpu_run);
+
+	smp_call_function_many(&cpus, ack_kick, (void*) cancel_cpu_run, wait_for_cpus);
```

Comment:
In the case where wait_for_cpus is false, this is wasteful--it requires waiting for any previous IPI to have finished first.

I would do something like:

```c
if (wait_for_cpus) {
    smp_call_function_many(...)
} else {
    if (cancel_cpu_run) {
      for (cpu in cpus) {
          WRITE_ONCE(mshv_vtl_the_run(cpu)->cancel, 1);
      }
    }
    __apic_send_IPI_mask(&cpus, RESCHEDULE_VECTOR);
    
}
```

Reply:
hmm i see, let me give that a try.

Reply:
Also, if we want the same on arm64, there's no ipi cpu mask equivalent that is usable by drivers, at least `__ipi_send_mask` and `smp_cross_call` don't seemed to be used by any drivers

Reply:
I've made it so we do this fast path on x64, but on other paths we just do smp_call_function_many (because there doesn't exist a good way to send a broadcast reschedule IPI). On ARM64, there's some hidden functions that aren't exposed to drivers.